pkg/storegateway/bucket.go

```diff
@@ -615,102 +621,308 @@ func (s *BucketStore) Series(req *storepb.SeriesRequest, srv storepb.Store_Serie
-		var (
-			iterationBegin = time.Now()
-			encodeDuration = time.Duration(0)
-			sendDuration   = time.Duration(0)
-			seriesCount    int
-			chunksCount    int
-		)
-
-		// Once the iteration is done we will update the stats.
-		defer stats.update(func(stats *queryStats) {
-			stats.mergedSeriesCount += seriesCount
-			stats.mergedChunksCount += chunksCount
-
-			// The time spent iterating over the series set is the
-			// actual time spent fetching series and chunks, encoding and sending them to the client.
-			// We split the timings to have a better view over how time is spent.
-			stats.streamingSeriesFetchSeriesAndChunksDuration += stats.streamingSeriesWaitBatchLoadedDuration
-			stats.streamingSeriesEncodeResponseDuration += encodeDuration
-			stats.streamingSeriesSendResponseDuration += sendDuration
-			stats.streamingSeriesOtherDuration += time.Duration(util_math.Max(0, int64(time.Since(iterationBegin)-
-				stats.streamingSeriesFetchSeriesAndChunksDuration-encodeDuration-sendDuration)))
-		})
+	spanName := "bucket_store_merge_all"
+	if req.StreamingChunksBatchSize > 0 {
+		spanName = "bucket_store_streaming_chunks_merge_all"
+	}
+	tracing.DoWithSpan(ctx, spanName, func(ctx context.Context, _ tracing.Span) {
+		err = s.sendSeriesChunks(req, srv, seriesSet, stats, iterationBegin)
+	})
+	if err != nil {
+		return
+	}
 
-		for seriesSet.Next() {
-			var lset labels.Labels
-			var series storepb.Series
+	if req.StreamingChunksBatchSize == 0 || req.SkipChunks {
+		// Hints and stats were not sent before, so send it now.
+		return s.sendHintsAndStats(srv, resHints, stats)
+	}
 
-			seriesCount++
+	return nil
+}
 
-			// IMPORTANT: do not retain the memory returned by seriesSet.At() beyond this loop cycle
-			// because the subsequent call to seriesSet.Next() may release it.
-			if req.SkipChunks {
-				lset, _ = seriesSet.At()
-			} else {
-				lset, series.Chunks = seriesSet.At()
+// sendStreamingSeriesLabelsHintsStats sends the labels of the streaming series.
+// Since hints and stats need to be sent before the "end of stream" streaming series message,
+// this function also sends the hints and the stats.
+func (s *BucketStore) sendStreamingSeriesLabelsHintsStats(
+	req *storepb.SeriesRequest,
+	srv storepb.Store_SeriesServer,
+	stats *safeQueryStats,
+	seriesSet storepb.SeriesSet,
+	resHints *hintspb.SeriesResponseHints,
+) error {
+	var (
+		encodeDuration = time.Duration(0)
+		sendDuration   = time.Duration(0)
+	)
+	// Once the iteration is done we will update the stats.
+	defer stats.update(func(stats *queryStats) {
+		// The time spent iterating over the series set is the
+		// actual time spent fetching series and chunks, encoding and sending them to the client.
+		// We split the timings to have a better view over how time is spent.
+		stats.streamingSeriesEncodeResponseDuration += encodeDuration
+		stats.streamingSeriesSendResponseDuration += sendDuration
```

Comment:
I will update the `other` part, but I don't think we should be doing the `stats.streamingSeriesFetchSeriesAndChunksDuration += stats.streamingSeriesWaitBatchLoadedDuration` here, because as I see `stats.streamingSeriesWaitBatchLoadedDuration` will include the cumulative duration for both streaming series and chunks, so doing it once at the end is correct unless I am reading it wrong. I will add a comment about it.

Reply:
The idea behind `streamingSeriesFetchSeriesAndChunksDuration` is to give an idea of how much time is spent fetching series or chunks. This duration is used to construct the histogram below. The idea is that this histogram adds up to the full request time. https://github.com/grafana/mimir/blob/bb7db110dcc32b533b58f6bb07be612dc52c80c2/pkg/storegateway/bucket.go#L902-L905 

Since we spend the time to fetch series twice and then spend some time to fetch chunks I think it's not a problem to record these twice.

Reply:
The latest refactor makes this case simpler where the timer is separated for fetching series and chunks (earlier the total duration was for sending series+chunks). It should be fixed now.